FileSystemAlreadyExistsException when recordBanner is run in parallel

Some of us use [mvnd](https://github.com/gnodet/mvnd) to build Camel Quarkus. Mvnd builds the modules in parallel by default. The recently added `BannerProcessor.recordBanner()` does not seem to be prepared for that because it throws the following exception when an application is built with mvnd:

```
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal [32mio.quarkus:quarkus-maven-plugin:1.3.0.CR1:build[m [1m(default)[m on project [36mcamel-quarkus-integration-test-stream[m: [1;31mFailed to build quarkus application[m
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:215)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
	at org.jboss.fuse.mvnd.builder.SmartBuilderImpl.buildProject(SmartBuilderImpl.java:176)
	at org.jboss.fuse.mvnd.builder.SmartBuilderImpl$ProjectBuildTask.run(SmartBuilderImpl.java:196)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.maven.plugin.MojoExecutionException: Failed to build quarkus application
	at io.quarkus.maven.BuildMojo.execute(BuildMojo.java:185)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
	... 10 common frames omitted
Caused by: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.BannerProcessor#recordBanner threw an exception: java.nio.file.FileSystemAlreadyExistsException
	at com.sun.nio.zipfs.ZipFileSystemProvider.newFileSystem(ZipFileSystemProvider.java:113)
	at java.nio.file.FileSystems.newFileSystem(FileSystems.java:326)
	at java.nio.file.FileSystems.newFileSystem(FileSystems.java:276)
	at io.quarkus.deployment.steps.BannerProcessor.isQuarkusCoreBanner(BannerProcessor.java:111)
	at io.quarkus.deployment.steps.BannerProcessor.getBanner(BannerProcessor.java:93)
	at io.quarkus.deployment.steps.BannerProcessor.readBannerFile(BannerProcessor.java:53)
	at io.quarkus.deployment.steps.BannerProcessor.recordBanner(BannerProcessor.java:42)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:938)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:273)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2027)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1551)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1442)
	at java.lang.Thread.run(Thread.java:748)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)

	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:206)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createProductionApplication(AugmentActionImpl.java:77)
	at io.quarkus.maven.BuildMojo.execute(BuildMojo.java:172)
	... 12 common frames omitted
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.BannerProcessor#recordBanner threw an exception: java.nio.file.FileSystemAlreadyExistsException
	at com.sun.nio.zipfs.ZipFileSystemProvider.newFileSystem(ZipFileSystemProvider.java:113)
	at java.nio.file.FileSystems.newFileSystem(FileSystems.java:326)
	at java.nio.file.FileSystems.newFileSystem(FileSystems.java:276)
	at io.quarkus.deployment.steps.BannerProcessor.isQuarkusCoreBanner(BannerProcessor.java:111)
	at io.quarkus.deployment.steps.BannerProcessor.getBanner(BannerProcessor.java:93)
	at io.quarkus.deployment.steps.BannerProcessor.readBannerFile(BannerProcessor.java:53)
	at io.quarkus.deployment.steps.BannerProcessor.recordBanner(BannerProcessor.java:42)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:938)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:273)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2027)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1551)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1442)
	at java.lang.Thread.run(Thread.java:748)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)

	at io.quarkus.builder.Execution.run(Execution.java:115)
	at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:135)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:204)
	... 14 common frames omitted
Caused by: java.nio.file.FileSystemAlreadyExistsException: null
	at com.sun.nio.zipfs.ZipFileSystemProvider.newFileSystem(ZipFileSystemProvider.java:113)
	at java.nio.file.FileSystems.newFileSystem(FileSystems.java:326)
	at java.nio.file.FileSystems.newFileSystem(FileSystems.java:276)
	at io.quarkus.deployment.steps.BannerProcessor.isQuarkusCoreBanner(BannerProcessor.java:111)
	at io.quarkus.deployment.steps.BannerProcessor.getBanner(BannerProcessor.java:93)
	at io.quarkus.deployment.steps.BannerProcessor.readBannerFile(BannerProcessor.java:53)
	at io.quarkus.deployment.steps.BannerProcessor.recordBanner(BannerProcessor.java:42)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:938)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:273)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2027)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1551)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1442)
	at java.lang.Thread.run(Thread.java:748)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

Steps to reproduce:

1. clone, build and add mvnd to PATH as described on https://github.com/gnodet/mvnd
2. Build recent Camel Quarkus (based on Quarkus 1.3.0.CR1) with mvnd:

```
git clone https://github.com/apache/camel-quarkus.git
cd camel-quarkus
mvnd clean install -DskipTests -e
```

Alternatively, this issue can be reproduced with stock Maven using `-T C1` in Camel Quarkus:

```
mvn clean install -DskipTests -e -T C1
```

When I run the above Maven command in Quarkus, I am getting a different unrelated error.

cc @geoand

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/BannerProcessor.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/BannerProcessor.java
@@ -2,20 +2,16 @@ package io.quarkus.deployment.steps;
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.net.URI;
 import java.net.URISyntaxException;
 import java.net.URL;
 import java.nio.charset.StandardCharsets;
-import java.nio.file.FileSystem;
-import java.nio.file.FileSystems;
-import java.nio.file.Files;
+import java.nio.file.Paths;
 import java.util.AbstractMap;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
 import java.util.Enumeration;
-import java.util.List;
 import java.util.Map;
 import java.util.Scanner;
+import java.util.jar.JarFile;
 
 import org.jboss.logging.Logger;
 
@@ -105,16 +101,14 @@ public class BannerProcessor {
             return false;
         }
 
+        String thisClassName = this.getClass().getName();
+        String jarPath = url.getPath().substring(0, url.getPath().lastIndexOf('!'));
+
         // We determine whether the banner is the default by checking to see if the jar that contains it also
         // contains this class. This way although somewhat complicated guarantees that any rename of artifacts
         // won't affect the check
-        try (FileSystem fileSystem = FileSystems.newFileSystem(url.toURI(), Collections.emptyMap())) {
-            String thisClassName = this.getClass().getName();
-            String[] parts = thisClassName.split("\\.");
-            List<String> rest = new ArrayList<>(parts.length - 1);
-            rest.addAll(Arrays.asList(parts).subList(1, parts.length - 1));
-            rest.add(parts[parts.length - 1] + ".class");
-            return Files.exists(fileSystem.getPath(parts[0], rest.toArray(new String[] {})));
+        try (JarFile jarFile = new JarFile(Paths.get(new URI(jarPath)).toFile())) {
+            return jarFile.getJarEntry(thisClassName.replace('.', '/') + ".class") != null;
         } catch (URISyntaxException e) {
             return false;
         }

